### DISCUSSION TITLE
Params Fail to Evaluate In Dag Body

### DISCUSSION THREAD

**Discussion by oliver-helix** [apache/airflow#39692] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.7.2

### What happened?

When providing airflow params, the parameter is not available in the body of the airflow dag.

### What you think should happen instead?

The value should be available in the airflow dag

### How to reproduce

```python
from airflow import DAG
from airflow.models.param import Param
from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator

with DAG("generate_password_dag", params={"test": Param("defeault", type="string")}) as dag:
    sql_operator = SQLExecuteQueryOperator(
        task_id="sql_operator",
        sql="sql/{{ params.test }}.sql",
        database="default",
    )
```

### Operating System

NA

### Versions of Apache Airflow Providers

NA

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

NA

### Anything else?

NA

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by Taragolis** [comment]
This rendered because you have `.sql` suffix in the end of your string. In this case it will tried to read template from the template file without rendering file path

**Comment by oliver-helix** [comment]
Thanks for your response @Taragolis! For my use-case, I'd like to determine the SQL file with a param provided in the UI

**Comment by josh-fell** [comment, marked as answer]
You can use [Jinja's `include` tag](https://jinja.palletsprojects.com/en/3.1.x/templates/#include) to render a file's contents.

For example, using this SQL file;
```sql
# dags/sql/default.sql
SELECT 'from a templated file';
```
and this task definition:
```py
from airflow.models.dag import DAG
from airflow.models.param import Param
from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator

with DAG(
    "generate_password_dag",
    params={"test": Param("default", type="string")},
):
    sql_operator = SQLExecuteQueryOperator(
        task_id="sql_operator",
        sql="{% include 'sql/' ~ params.test ~ '.sql' %}",
        conn_id="postgres",
    )
```
In the above task definition, `~` is string concatenation in Jinja.

Task log:
<img width="941" alt="image" src="https://github.com/apache/airflow/assets/48934154/3ddf0c00-c10a-40fa-97f0-b2993bc26bc9">

@oliver-helix Can you try that approach and see if it works please?

**Comment by oliver-helix** [comment]
Thank you @josh-fell! This indeed works. Curious why `sql="{% include 'sql/' ~ params.test ~ '.sql' %}"` works when `sql="sql/{{ params.test }}.sql"` does not?

**Comment by Taragolis** [comment]
> works when sql="sql/{{ params.test }}.sql" does not?

It doesn't work by design, see: https://github.com/apache/airflow/issues/39651#issuecomment-2113385429

> Curious why sql="{% include 'sql/' ~ params.test ~ '.sql' %}" works 

It is just a Jinja statement which rendered

**Comment by oliver-helix** [comment]
@Taragolis @josh-fell

> TL;DR: How can I force template render before the operator execution?

Thank you two for helping earlier. I've reached another road block and was hoping you two could help again 😅

Struggling to render joined params. In the following example, I've tried two ways, using your earlier `{% include ...` logic, and using python format string as recommended in the docs. However, in both attempts, airflow injects the raw string value into the SQL query. 

Error:
```
# with name = "{% include params.site ~ '_' ~ params.submission %}"
LINE 9: DROP SCHEMA IF EXISTS {% include params.site ~ '_' ~ par...
```
or 
```
# with name = "{{ params.site }}_{{ params.submission }}"
LINE 9: DROP SCHEMA IF EXISTS {{ params.site }}_{{ params.submis...
```

Code to reproduce the problem:

```py
from datetime import datetime

from airflow.models import Param, DAG
from airflow.providers.postgres.operators.postgres import PostgresOperator


with DAG(
    dag_id="my_dag",
    start_date=datetime(2022, 6, 1),
    params={
        "site": Param("mysite", type="string"),
        "submission": Param("mysubmission", type="string"),
    },
) as dag:
    name = "{{ params.site }}_{{ params.submission }}"
    # name = "{% include params.site ~ '_' ~ params.submission %}"

    task = PostgresOperator(
        task_id="rename_previous_submission_schemas",
        postgres_conn_id="my-conn-id",
        sql="{% include 'sql/' ~ params.site ~ '.sql' %}",
        params={"name": name},
    )
```
